packages/vite/src/shared/moduleRunnerTransport.ts

```diff
@@ -0,0 +1,287 @@
+import { nanoid } from 'nanoid/non-secure'
+import type { CustomPayload, HotPayload } from 'types/hmrPayload'
+import { promiseWithResolvers } from './utils'
+
+export type ModuleRunnerTransportHandlers = {
+  onMessage: (data: HotPayload) => void
+  onDisconnection: () => void
+}
+
+/**
+ * "send and connect" or "invoke" must be implemented
+ */
+export interface ModuleRunnerTransport {
+  connect?(handlers: ModuleRunnerTransportHandlers): Promise<void> | void
+  disconnect?(): Promise<void> | void
+  send?(data: HotPayload): Promise<void> | void
+  invoke?(
+    data: HotPayload,
+  ): Promise<{ /** result */ r: any } | { /** error */ e: any }>
+  timeout?: number
+}
+
+type InvokeableModuleRunnerTransport = Omit<ModuleRunnerTransport, 'invoke'> &
+  Required<Pick<ModuleRunnerTransport, 'send'>> & {
+    invoke(name: string, data: any): any
+  }
+
+const createInvokeableTransport = (
+  transport: ModuleRunnerTransport,
+): InvokeableModuleRunnerTransport => {
+  if (transport.invoke) {
+    const sendOrInvoke = transport.send ?? transport.invoke
+    return {
+      ...transport,
+      async send(data) {
+        await sendOrInvoke(data)
+      },
+      async invoke(name, data) {
+        const result = await transport.invoke!({
+          type: 'custom',
+          event: name,
+          invoke: 'send',
+          data,
+        } satisfies CustomPayload)
+        if ('e' in result) {
+          throw result.e
+        }
+        return result.r
+      },
+    }
+  }
+
+  if (!transport.send || !transport.connect) {
+    throw new Error(
+      'transport must implement send and connect when invoke is not implemented',
+    )
+  }
+
+  const rpcPromises = new Map<
+    string,
+    {
+      resolve: (data: any) => void
+      reject: (data: any) => void
+      timeoutId?: ReturnType<typeof setTimeout>
+    }
+  >()
+
+  return {
+    ...transport,
+    connect({ onMessage, onDisconnection }) {
+      transport.connect!({
+        onMessage(data) {
+          if (
+            data.type === 'custom' &&
+            data.invoke &&
+            data.invoke.startsWith('response:')
+          ) {
+            const invokeId = data.invoke.slice('response:'.length)
+            const promise = rpcPromises.get(invokeId)
+            if (!promise) return
+
+            if (promise.timeoutId) clearTimeout(promise.timeoutId)
+
+            rpcPromises.delete(invokeId)
+
+            const { e, r } = data.data
+            if (e) {
+              promise.reject(e)
+            } else {
+              promise.resolve(r)
+            }
+            return
+          }
+          onMessage(data)
+        },
+        onDisconnection,
+      })
+    },
+    disconnect() {
+      rpcPromises.forEach((promise) => {
+        promise.reject(new Error('transport was disconnected'))
```

Comment:
can we print more information? at least the name of the method. if it's unhandled, it will be hard to track (like here https://github.com/antfu/birpc/pull/22/files)

Reply:
Makes sense 👍  Added in 38974274105ea956c90fa2c431f374bb5ff6d53c